File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
ListNode.__lt__ = lambda self, other: self.val < other.val

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return

        cur = res = ListNode()
        min_heap = []

        for head in lists:
            if head:
                heapq.heappush(min_heap, head)

        while min_heap:
            cur.next = (node := heapq.heappop(min_heap))
            cur = cur.next

            if node.next:
                heapq.heappush(min_heap, node.next)

        return res.next

File: Data Structures & Algorithms/min-cost-to-connect-points/submission-1.cpp
class Solution {
   public:
    int minCostConnectPoints(vector<vector<int>>& points) {
        int n = points.size(), node = 0;
        vector<int> dist(n, 100000000);
        vector<bool> visit(n);
        int edges = 0, res = 0;

        while (edges < n - 1) {
            visit[node] = true;
            int next = -1;

            for (int i = 0; i < n; ++i) {
                if (visit[i]) continue;
                int cur = abs(points[i][0] - points[node][0]) + abs(points[i][1] - points[node][1]);
                dist[i] = min(dist[i], cur);

                if (next == -1 || dist[i] < dist[next]) {
                    next = i;
                }
            }
            res += dist[next];
            node = next;
            ++edges;
        }
        return res;
    }
};